Error while checking build/O2/fullCI for f6a89270cf0960453a4bc4d2cbd6546ad3866da1 at 2024-07-23 17:46:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
/sw/SOURCES/O2/13318-slc8_x86-64/0/Framework/Core/src/DataProcessingDevice.cxx:2305:28: error: 'Timeslice' is not a member of 'o2::framework::CompletionPolicy'
/sw/SOURCES/O2/13318-slc8_x86-64/0/Framework/Core/src/DataProcessingDevice.cxx:2308:28: error: 'Slot' is not a member of 'o2::framework::CompletionPolicy'
/sw/SOURCES/O2/13318-slc8_x86-64/0/Framework/Core/src/DataProcessingDevice.cxx:2311:28: error: 'Any' is not a member of 'o2::framework::CompletionPolicy'
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13318/f6a89270cf0960453a4bc4d2cbd6546ad3866da1/build_O2_fullCI/pretty.html).

--- a/Framework/Core/include/Framework/CompletionPolicy.h
+++ b/Framework/Core/include/Framework/CompletionPolicy.h
@@ -62,6 +62,13 @@ struct CompletionPolicy {
     Retry,
   };
 
+  /// Order in which the completed slots must be consumed
+  enum struct CompletionOrder {
+    Any,
+    Timeslice,
+    Slot
+  };
+
   using Matcher = std::function<bool(DeviceSpec const& device)>;
   using InputSetElement = DataRef;
   using CallbackFull = std::function<CompletionOp(InputSpan const&, std::vector<InputSpec> const&, ServiceRegistryRef&)>;
@@ -90,6 +97,8 @@ struct CompletionPolicy {
   /// not needed if the policy always happens to consume / discard
   /// data.
   bool balanceChannels = true;
+
+  CompletionOrder order = CompletionOrder::Any;
 
   /// Helper to create the default configuration.
   static std::vector<CompletionPolicy> createDefaultPolicies();

--- a/Framework/Core/src/DataProcessingDevice.cxx
+++ b/Framework/Core/src/DataProcessingDevice.cxx
@@ -2301,6 +2301,17 @@ bool DataProcessingDevice::tryDispatchComputation(ServiceRegistryRef ref, std::v
   using namespace o2::framework;
   stats.updateStats({(int)ProcessingStatsId::PENDING_INPUTS, DataProcessingStats::Op::Set, static_cast<int64_t>(relayer.getParallelTimeslices() - completed.size())});
   stats.updateStats({(int)ProcessingStatsId::INCOMPLETE_INPUTS, DataProcessingStats::Op::Set, completed.empty() ? 1 : 0});
+  switch (spec.completionPolicy.order) {
+    case CompletionPolicy::CompletionOrder::Timeslice:
+      std::sort(completed.begin(), completed.end(), [](auto const& a, auto const& b) { return a.timeslice.value < b.timeslice.value; });
+      break;
+    case CompletionPolicy::CompletionOrder::Slot:
+      std::sort(completed.begin(), completed.end(), [](auto const& a, auto const& b) { return a.slot.index < b.slot.index; });
+      break;
+    case CompletionPolicy::CompletionOrder::Any:
+    default:
+      break;
+  }
 
   for (auto action : completed) {
     O2_SIGNPOST_ID_GENERATE(aid, device);

